### PR TITLE
[READY]The smell of death! New machine to puts the spectre in spectroscopy.

### PR DESCRIPTION
**Pull request by Krysonism**
## About The Pull Request

This PR adds a new machine; the ectoscopic sniffer, to the robotics lab on each map.

The sniffer is small desktop machine used for detecting paranormal activity, it displays a green flash and a chime every time it is touched(as in attacked) by a ghost.

It has a single wire inside, when this wire is pulsed, it triggers the animation and chime, as if touched by a ghost, maybe some enterprising crewmember can use this to their advantage. When this wire is cut, the device won't trigger, no matter how many spooks or haunts touch it.

ectoscopic_sniffer

This PR is heavily inspired by this previous PR: https://github.com/tgstation/tgstation/pull/58840

## Why It's Good For The Game
This PR is above all fun. 

Of course, it can be used as the author of the previous PR intended; to alert the roboticists of souls willing to occupy a positronic brain.

Clever spacemen will always figure out how to use this kind of device for more creative parascientifc research. 

It is still limited by its low bitrate and the very low quality of the information as compared to something like ghost court, since a single malevolent spirit is able to feed the living bullshit.

## Changelog
:cl:
add: Added a new machine to the robotics lab; the ectoscopic sniffer.
/:cl:

### CODE REVIEW

File: code/game/machinery/ecto_sniffer.dm
Comment (Mothblocks) [inline code/game/machinery/ecto_sniffer.dm:52]: `wrench_act`

File: code/game/machinery/ecto_sniffer.dm
Comment (Mothblocks) [inline code/game/machinery/ecto_sniffer.dm:54]:
```suggestion
		to_chat(user, "<span class ='notice'>You [anchored ? "unanchor" : "anchor"] [src].")
```

File: code/game/machinery/ecto_sniffer.dm
Comment (Mothblocks) [inline code/game/machinery/ecto_sniffer.dm:18]:
```suggestion
	if(user in ghosts_sampled) //anti-spam protection, also helps limit the bitrate.
```

`locate` shouldn't be necessary here. Also, don't use tabs inline. Just a space.

File: code/game/machinery/ecto_sniffer.dm
Comment (Mothblocks) [inline code/game/machinery/ecto_sniffer.dm:60]: I'm mixed on this being a to_chat (see https://hackmd.io/@tgstation/SJIle5wYd), but at a 15 second delay I can hesitantly let this slide.

```suggestion
	to_chat(user, "<span class='nicegreen'>The coating of ectoplasmic residue you left on [src]'s sensors has decayed.</span>")
```

File: code/game/machinery/ecto_sniffer.dm
Comment (Mothblocks) [inline code/game/machinery/ecto_sniffer.dm:14]: Clear this on Destroy.

File: code/game/machinery/ecto_sniffer.dm
Comment (Mothblocks) [inline code/game/machinery/ecto_sniffer.dm:32]:
```suggestion
		addtimer(CALLBACK(src, .proc/clear_residue, activator), 15 SECONDS)
```

File: code/datums/wires/ecto_sniffer.dm
Comment (Mothblocks) [inline code/datums/wires/ecto_sniffer.dm:17]:
```suggestion
	our_sniffer.sensor_enabled = mend
```

**Review by Mothblocks — CHANGES_REQUESTED** [review]

File: code/game/machinery/ecto_sniffer.dm
Comment (Mothblocks) [reply code/game/machinery/ecto_sniffer.dm:14]: Also, these should be weakrefs so a delete of the ghost doesn't hard-del.

Comment (Krysonism) [reply code/game/machinery/ecto_sniffer.dm:14]: I have nulled the list on destroy.

I'd love to do the weakref part but i don't know how. Maybe i can figure it out tomorrow.

Comment (LemonInTheDark) [reply code/game/machinery/ecto_sniffer.dm:14]: yeah this really should be weakrefs

File: code/game/machinery/ecto_sniffer.dm
Comment (ATH1909) [inline code/game/machinery/ecto_sniffer.dm:3]: false advertising, this doesn't react to the ectoplasm item(s).

Comment (norill) [reply code/game/machinery/ecto_sniffer.dm:3]: just make it say what does it do, no one's got time for bullshit riddles

**Comment by Time-Green** [comment]
Could you add a second wire that sends a signal when the machine is pulsed? That way you can do stupid shit like let ghosts bolt doors and trigger bombs ~~and give ghosts the ability to controle integrated circuit drones~~

**Comment by norill** [comment]
> 
> 
> Could you add a second wire that sends a signal when the machine is pulsed? That way you can do stupid shit like let ghosts bolt doors and trigger bombs ~and give ghosts the ability to controle integrated circuit drones~

wires cant send signals iirc

**Comment by Krysonism** [comment]
> 
> 
> Could you add a second wire that sends a signal when the machine is pulsed? That way you can do stupid shit like let ghosts bolt doors and trigger bombs ~and give ghosts the ability to controle integrated circuit drones~

I considered adding USB integration in the future. This might quickly become too cancerous and lead to removal though.

File: code/game/machinery/ecto_sniffer.dm
Comment (Krysonism) [reply code/game/machinery/ecto_sniffer.dm:3]: I think  the primary job of the description is to add a little flavour. 

If after reading the description and observing the device, they still don't get what it does, they should be asking their science department coworkers. 

Learning from other people ingame is always more rewarding, fosters interactions and often allows you to pick up some neat tricks.

**Comment by Dorsisdwarf** [comment]
The return of chairspin

File: code/game/machinery/ecto_sniffer.dm
Comment (norill) [reply code/game/machinery/ecto_sniffer.dm:3]: flavour is ok, but it also rises the already enormous entry barrier. you have literally thousands of items on the station and guessing what they do and how to use them is very time-consuming. if anything, cryptic, non-informative descriptions lead to code-diving

**Comment by Mothblocks** [comment]
add usb integration to all the things ✨

File: code/game/machinery/ecto_sniffer.dm
Comment (Mothblocks) [inline code/game/machinery/ecto_sniffer.dm:74]: You don't have to resolve, you can just `IS_WEAKREF_OF`

File: code/game/machinery/ecto_sniffer.dm
Comment (Mothblocks) [inline code/game/machinery/ecto_sniffer.dm:72]: `as anything`

File: code/game/machinery/ecto_sniffer.dm
Comment (Mothblocks) [inline code/game/machinery/ecto_sniffer.dm:27]: Use `IS_WEAKREF_OF`

File: code/game/machinery/ecto_sniffer.dm
Comment (Mothblocks) [inline code/game/machinery/ecto_sniffer.dm:42]:
```suggestion
	balloon_alert(user, "sniffer turned [on ? "off" : "on"]")
```

File: code/game/machinery/ecto_sniffer.dm
Comment (Mothblocks) [inline code/game/machinery/ecto_sniffer.dm:58]:
```suggestion
	balloon_alert(user, "sniffer [anchored ? "unanchor" : "anchor"]")
```

**Review by Mothblocks — CHANGES_REQUESTED** [review]
Your weakrefs look good, just some minor concerns, also balloon alerts are prefered instead of `to_chat` where archival is not necessary.

**Comment by PKPenguin321** [comment]
Sounds cool. Will invariably need a policy thread in the future when someone asks ghosts to "beep once if this guy is valid" then kills an innocent person over it, but we can get to that when we need to

File: code/game/machinery/ecto_sniffer.dm
Comment (Fikou) [inline code/game/machinery/ecto_sniffer.dm:58]: shouldn't this be unanchored/anchored?

Comment (Mothblocks) [reply code/game/machinery/ecto_sniffer.dm:58]: `set_anchored(!anchored)` is after this (though I never quite liked the pattern of setting the variable *after* displaying it)

Comment (Fikou) [reply code/game/machinery/ecto_sniffer.dm]: i meant from grammar side, should be sniffer anchored instead of sniffer anchor

**Review by Twaticus — APPROVED** [review]

File: code/game/machinery/ecto_sniffer.dm
Comment (Mothblocks) [reply code/game/machinery/ecto_sniffer.dm]: oh ya ur right

File: code/game/machinery/ecto_sniffer.dm
Comment (coiax) [inline code/game/machinery/ecto_sniffer.dm:36]: Can't we also have the argument passed to the timer system a weakrefence to the ghost as well?

File: code/game/machinery/ecto_sniffer.dm
Comment (coiax) [inline code/game/machinery/ecto_sniffer.dm:27]: Doesn't this mean we'll call super() twice in this case?

Also, how do we feel about giving the clicking ghost some feedback that it's too soon?

File: code/game/machinery/ecto_sniffer.dm
Comment (coiax) [inline code/game/machinery/ecto_sniffer.dm:70]:
```suggestion

	ectoplasmic_residues = null
	. = ..()
```

Destroy cleanup before calling parent.

File: code/game/machinery/ecto_sniffer.dm
Comment (coiax) [inline code/game/machinery/ecto_sniffer.dm:79]: Could provide a clickable link to take you to the machine again?

**Comment by coiax** [comment]
Additional points:

1) Does the existence of this machine imply that positronic brains are powered by the souls of the dead? We know, that mechanically that is true, but I always thought the lore implications was that positronic brains just... sometimes didn't work because they were emergent AI?

2) Is the machine explicitly non-constructible destructible, or is that an oversight? I don't tend to like machines that can't be built in universe.

**Comment by Krysonism** [comment]
> 
> 
> Additional points:
> 
>     1. Does the existence of this machine imply that positronic brains are powered by the souls of the dead?

No. I noticed this lore inconsistency in the positronic alert console pr. My device doesn't mention positronic technology anywhere. I imagine the lore reason for the device being in the robotics lab has more to do with to do with preventing botched borgings and such.

>     2. Is the machine explicitly non-constructible destructible, or is that an oversight? I don't tend to like machines that can't be built in universe.

I have considered making it constructable, as long as it is not printable. I think the tech loot spawner that is used in several space ruins could be a good place for the board.

**Comment by coiax** [comment]
> I have considered making it constructable, as long as it is not printable. I think the tech loot spawner that is used in several space ruins could be a good place for the board.

Alternatively, the machine recipe requires some sort of unique item, like "refined ectoplasm". Print all the boards you want, but it's this "quantum ghost juice" that makes the machine go brrrrrr. Either way works.

**Comment by JJRcop** [comment]
I think all ghosts should be alerted when this is used, with a JMP ref, so they can meddle with any attempt at intelligent communication if they wish. Release the chaos.

**Comment by Krysonism** [comment]
I am going to implement the changes the maintainers requested tomorrow.

> 
> 
> I think all ghosts should be alerted when this is used, with a JMP ref, so they can meddle with any attempt at intelligent communication if they wish. Release the chaos.

I think it would be funnier if i leave it up to the ghost who first finds the machine to choose between calling out the location in ghost chat vs keeping quiet to try to establish a more personal line of communication.

File: code/game/machinery/ecto_sniffer.dm
Comment (Krysonism) [reply code/game/machinery/ecto_sniffer.dm:79]: I don't know how to do that.

File: code/game/machinery/ecto_sniffer.dm
Comment (Krysonism) [reply code/game/machinery/ecto_sniffer.dm:70]: ok

File: code/game/machinery/ecto_sniffer.dm
Comment (Krysonism) [reply code/game/machinery/ecto_sniffer.dm]: ok

File: code/game/machinery/ecto_sniffer.dm
Comment (Krysonism) [reply code/game/machinery/ecto_sniffer.dm]: ok

File: code/game/machinery/ecto_sniffer.dm
Comment (JJRcop) [reply code/game/machinery/ecto_sniffer.dm:79]: Look at stuff that does it, like a singularity being created.

Comment (JJRcop) [reply code/game/machinery/ecto_sniffer.dm:79]: Was a lot harder to find than I thought. https://github.com/tgstation/tgstation/blob/master/code/__DEFINES/say.dm#L78

File: code/game/machinery/ecto_sniffer.dm
Comment (JJRcop) [inline code/game/machinery/ecto_sniffer.dm:71]: ..()

File: code/game/machinery/ecto_sniffer.dm
Comment (JJRcop) [reply code/game/machinery/ecto_sniffer.dm]: while outdated, this review was not resolved

**Comment by Greniza** [comment]
Why the Robotics lab? Why not, say, the Curator office? Or Chaplain's office? They deal with paranormal measures far more than Robotics does.

**Comment by JJRcop** [comment]
@Greniza 

>This PR is heavily inspired by this previous PR: #58840

File: code/game/machinery/ecto_sniffer.dm
Comment (Krysonism) [reply code/game/machinery/ecto_sniffer.dm]: I have added a link now with the help of JJRcop.

File: code/game/machinery/ecto_sniffer.dm
Comment (Krysonism) [reply code/game/machinery/ecto_sniffer.dm]: done

**Review by coiax — CHANGES_REQUESTED** [review]
Just checked a suspicion. It's possible, if you were a regular dead person, to hop outside your body as a ghost, hit the sniffer, then Renter Your Body, and leave again.

That will create a totally new ghost, bypassing the reference check.

A possible solution is to instead treat the ckeys of people who use it, and then remove those after the X seconds are up, and only sending them the "HEY YOU CAN USE THIS NOW" if they're still an observer?

File: code/game/machinery/ecto_sniffer.dm
Comment (coiax) [inline code/game/machinery/ecto_sniffer.dm:77]:
```suggestion
/obj/machinery/ecto_sniffer/proc/clear_residue(ghost_ckey)
```
Linter is angry because of this.

**Comment by coiax** [comment]
You think you could get a new sound for this? <3

**Pull request merged**
